File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/activity/overtime_controller.py
# app/Http/Controllers/activity/overtime_controller.py

import inspect
import logging
from django.contrib import messages
from django.core.exceptions import PermissionDenied
from django.db import transaction
from django.http import HttpRequest, HttpResponse, JsonResponse
from django.shortcuts import get_object_or_404, redirect, render
from ....Models.activity.overtime import Overtime
from ....Models.individuals.employee import Employee
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from .._helpers.http import get_redirect_url
from .._traits.controller import Controller

logger = logging.getLogger(__name__)

class OvertimeController(Controller):
  
  def index(self, request: HttpRequest) -> JsonResponse:
    try:
      try:
        self.authorize('update',['Overtime'])
        can_edit_all = True
      except PermissionDenied:
        can_edit_all = False
      # fetch and serialize
      ots = Overtime.objects.filter(created_by=request.user.creatorId())
      payload = [{
        'id': str(ot.id),
        'employee_id': str(ot.employee_id),
        'title': ot.title,
        'number_of_days': ot.number_of_days,
        'hours': ot.hours,
        'rate': ot.rate,
        'can_edit': can_edit_all
      } for ot in ots]
      return JsonResponse(payload, safe=False)
    except PermissionDenied as e:
      return default_permission_denial(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger,
        json={'error': 'Permission denied.'}
      )
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger,
        json={'error': str(e)}
      )  
  
  def overtime_create(self, request: HttpRequest, eid: str) -> HttpResponse:
    try:
      self.authorize('create',['Overtime'])
      emp = get_object_or_404(Employee, pk=eid)
      return render(request,'overtime/create.html',{'employee':emp})
    except PermissionDenied as e:
      return default_permission_denial(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

  def store(self, request: HttpRequest) -> HttpResponse:
    try:
      self.authorize('create',['Overtime'])
      data = request.POST
      missing = [f for f in ('employee_id','title','number_of_days','hours','rate') if not data.get(f)]
      if missing:
        messages.error(
          request,
          f"{missing[0].replace('_',' ').capitalize()} is required."
        )
        return redirect(get_redirect_url(request))
      with transaction.atomic():
        ot = Overtime(
          employee_id=data['employee_id'],
          title=data['title'],
          number_of_days=data['number_of_days'],
          hours=data['hours'],
          rate=data['rate'],
          created_by=request.user.creatorId()
        )
        ot.save()
      messages.success(request,'Overtime successfully created.')
      return redirect(get_redirect_url(request))
    except PermissionDenied as e:
      return default_permission_denial(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

  def show(self, request: HttpRequest, overtime_id: str) -> HttpResponse:
    try:
      # redirects to commission index
      return redirect('commission_index')
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

  def edit(self, request: HttpRequest, overtime_id: str) -> HttpResponse:
    try:
      self.authorize('update',['Overtime',overtime_id])
      ot = get_object_or_404(Overtime, pk=overtime_id)
      if ot.created_by != request.user.creatorId():
        messages.error(request,'Permission denied.')
        return redirect(get_redirect_url(request))
      return render(request,'overtime/edit.html',{'overtime':ot})
    except PermissionDenied as e:
      return default_permission_denial(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

  def update(self, request: HttpRequest, overtime_id: str) -> HttpResponse:
    try:
      self.authorize('update',['Overtime',overtime_id])
      ot = get_object_or_404(Overtime, pk=overtime_id)
      if ot.created_by != request.user.creatorId():
        messages.error(request,'Permission denied.')
        return redirect(get_redirect_url(request))
      data = request.POST
      missing = [f for f in ('title','number_of_days','hours','rate') if not data.get(f)]
      if missing:
        messages.error(
          request,
          f"{missing[0].replace('_',' ').capitalize()} is required."
        )
        return redirect(get_redirect_url(request))
      ot.title = data['title']
      ot.number_of_days = data['number_of_days']
      ot.hours = data['hours']
      ot.rate = data['rate']
      ot.save()
      messages.success(request,'Overtime successfully updated.')
      return redirect(get_redirect_url(request))
    except PermissionDenied as e:
      return default_permission_denial(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

  def destroy(self, request: HttpRequest, overtime_id: str) -> HttpResponse:
    try:
      self.authorize('delete',['Overtime',overtime_id])
      ot = get_object_or_404(Overtime, pk=overtime_id)
      if ot.created_by != request.user.creatorId():
        messages.error(request,'Permission denied.')
        return redirect(get_redirect_url(request))
      ot.delete()
      messages.success(request,'Overtime successfully deleted.')
      return redirect(get_redirect_url(request))
    except PermissionDenied as e:
      return default_permission_denial(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=f'{self.__class__.__name__}::{inspect.currentframe().f_code.co_name}',
        logger=logger
      )

File: _inc/laravel/app/Http/Controllers/Activity/OvertimeController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{DatabaseConstants, ViewsConstants as VW};
use App\Models\{Employee, Overtime};
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Http\{JsonResponse, RedirectResponse, Request};
use Illuminate\Support\Facades\{Log, Route, Validator, View as ViewFacade};
use Illuminate\View\View;

use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
use App\Traits\HasCrudConstants;
class OvertimeController extends Controller
{
    use HasCrudConstants;

  private const PERM_CREATE = 'create overtime';
  private const PERM_DELETE = 'delete overtime';
  private const PERM_EDIT = 'edit overtime';
  private const PERM_MANAGE = 'manage overtime';

  public function index(Request $request): View|RedirectResponse|JsonResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = VW::OVT . '.index';
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $action, $class, $viewPath) {
      Log::info("[{$class}::{$action}] start", ['creator_id' => $req->user()?->creatorId()]);
      if (!self::authorizePerm($req, self::PERM_MANAGE)) return redirect()->back();
      try {
        $creatorId = $req->user()->creatorId();
        $fetchStart = microtime(true);
        $overtimes = Overtime::where(DatabaseConstants::COL_TABLE_CREATOR, $creatorId)->get();
        $this->logExecutionTime($fetchStart, $action, 'fetchOvertimes');
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        Log::info("[{$class}::{$action}] complete", ['count' => is_countable($overtimes) ? count($overtimes) : null]);
        return view($viewPath, compact('overtimes'));
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'creator_id' => $req->user()?->creatorId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public const OVT_CRT = 'overtimeCreate';
  public function overtimeCreate(int|string $id): View|RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = VW::OVT . '.create';
    return $this->measureProfile($action, function () use ($id, $action, $class, $viewPath) {
      Log::info("[{$class}::{$action}] start", ['employee_id_param' => $id]);
      try {
        $findStart = microtime(true);
        $employee = Employee::find($id);
        $this->logExecutionTime($findStart, $action, 'findEmployee');
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        $renderStart = microtime(true);
        $resp = view($viewPath, compact('employee'));
        $this->logExecutionTime($renderStart, $action, 'renderOvertimeCreate');
        Log::info("[{$class}::{$action}] complete", ['has_employee' => (bool) $employee]);
        return $resp;
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'employee_id_param' => $id, 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        Log::error("[{$class}::{$action}] error", ['error' => $e->getMessage()]);
        throw $e;
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'employee_id_param' => $id]);
  }

  public function store(Request $request): RedirectResponse|JsonResponse|null
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['creator_id' => $req->user()?->creatorId(), 'input_keys' => array_keys($req->all())]);
      if (!self::authorizePerm($req, self::PERM_CREATE)) return redirect()->back();
      $rules = ['employee_id' => 'required', 'title' => 'required', 'number_of_days' => 'required', 'hours' => 'required', 'rate' => 'required'];
      $valStart = microtime(true);
      $validator = Validator::make($req->all(), $rules);
      $this->logExecutionTime($valStart, $action, 'buildValidator');
      if ($validator->fails()) return redirect()->back()->with('error', $validator->getMessageBag()->first());
      try {
        $overtime = new Overtime();
        foreach (['employee_id', 'title', 'number_of_days', 'hours', 'rate'] as $k) $overtime->$k = $req->input($k);
        $overtime->created_by = $req->user()->creatorId();
        $saveStart = microtime(true);
        $overtime->save();
        $this->logExecutionTime($saveStart, $action, 'saveOvertime');
        Log::info("[{$class}::{$action}] success", ['overtime_id' => $overtime->getKey()]);
        return redirect()->back()->with('success', __('Overtime successfully created.'));
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'creator_id' => $req->user()?->creatorId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public function show(Overtime $overtime): RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    return $this->measureProfile($action, function () use ($overtime, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['overtime_id' => $overtime->getKey()]);
      try {
        $redirStart = microtime(true);
        $resp = redirect()->route(VW::COM . '.index')->with($overtime);
        $this->logExecutionTime($redirStart, $action, 'redirectToIndex');
        Log::info("[{$class}::{$action}] complete", ['overtime_id' => $overtime->getKey()]);
        return $resp;
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'overtime_id' => $overtime->getKey(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        Log::error("[{$class}::{$action}] error", ['error' => $e->getMessage()]);
        throw $e;
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'overtime_id' => $overtime->getKey()]);
  }

  public function edit(Request $request, int|string $overtime): View|JsonResponse|RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = VW::OVT . '.edit';
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $overtime, $action, $class, $viewPath) {
      Log::info("[{$class}::{$action}] start", ['overtime_param' => $overtime, 'creator_id' => $req->user()?->creatorId()]);
      if (!self::authorizePerm($req, self::PERM_EDIT)) return response()->json(['error' => __('Permission denied.')], 401);
      try {
        $getStart = microtime(true);
        $ot = self::getOvertime($req, $overtime);
        $this->logExecutionTime($getStart, $action, 'getOvertime');
        if (!$ot) return response()->json(['error' => __('Permission denied.')], 401);
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        $renderStart = microtime(true);
        $resp = view($viewPath, compact('overtime'));
        $this->logExecutionTime($renderStart, $action, 'renderOvertimeEdit');
        Log::info("[{$class}::{$action}] complete", ['overtime_param' => $overtime]);
        return $resp;
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'overtime_param' => $overtime, 'creator_id' => $req->user()?->creatorId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'overtime_param' => $overtime]);
  }

  public function update(Request $request, int|string $overtime): RedirectResponse|JsonResponse|null
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $overtime, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['overtime_param' => $overtime, 'creator_id' => $req->user()?->creatorId(), 'input_keys' => array_keys($req->all())]);
      if (!self::authorizePerm($req, self::PERM_EDIT)) return redirect()->back();
      $ot = self::getOvertime($req, $overtime);
      if (!$ot) return redirect()->back();
      $rules = ['title' => 'required', 'number_of_days' => 'required', 'hours' => 'required', 'rate' => 'required'];
      $valStart = microtime(true);
      $validator = Validator::make($req->all(), $rules);
      $this->logExecutionTime($valStart, $action, 'buildValidator');
      if ($validator->fails()) return redirect()->back()->with('error', $validator->getMessageBag()->first());
      try {
        foreach (['title', 'number_of_days', 'hours', 'rate'] as $k) $ot->$k = $req->input($k);
        $saveStart = microtime(true);
        $ot->save();
        $this->logExecutionTime($saveStart, $action, 'saveOvertime');
        Log::info("[{$class}::{$action}] success", ['overtime_id' => $ot->getKey()]);
        return redirect()->back()->with('success', __('Overtime successfully updated.'));
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'overtime_param' => $overtime, 'creator_id' => $req->user()?->creatorId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'overtime_param' => $overtime]);
  }

  public function destroy(Request $request, Overtime $overtime): RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $overtime, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['overtime_id' => $overtime->getKey(), 'creator_id' => $req->user()?->creatorId()]);
      if (!self::authorizePerm($req, self::PERM_DELETE)) return redirect()->back();
      if ($overtime->created_by !== $req->user()->creatorId()) return defaultPermissionDenial($req, new AuthorizationException(), $class . '::' . $action);
      try {
        $delStart = microtime(true);
        $overtime->delete();
        $this->logExecutionTime($delStart, $action, 'deleteOvertime');
        Log::info("[{$class}::{$action}] success", ['overtime_id' => $overtime->getKey()]);
        return redirect()->back()->with('success', __('Overtime successfully deleted.'));
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'overtime_id' => $overtime->getKey(), 'creator_id' => $req->user()?->creatorId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'overtime_id' => $overtime->getKey()]);
  }

  protected static function authorizePerm(Request $request, string $perm): bool
  {
    if ($request->user()->can($perm)) return true;
    defaultPermissionDenial(
      $request,
      new AuthorizationException(),
      __CLASS__ . '::' . debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 1)[0]['function']
    );
    return false;
  }

  protected static function getOvertime(Request $request, int|string $id): ?Overtime
  {
    $ot = Overtime::find($id);
    if (!$ot || $ot->created_by !== $request->user()->creatorId()) {
      defaultPermissionDenial(
        $request,
        new AuthorizationException(),
        __CLASS__ . '::' . debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 1)[0]['function']
      );
      return null;
    }
    return $ot;
  }
}
